### PR TITLE
Initial Layers implementation

### PR DESCRIPTION
**Pull request by hearnadam**
/claim #203 
/split @kitlangton 

This PR creates a Layer implementation with memoization, combinators, and supports arbitary effects.

It's possible we can condense it into two type parameters, but that may need to be a followup exercise.

The remaining work involves developing a macro to combine layers automagically, but that will be a separate PR.

Written by @kitlangton and I.

### CODE REVIEW

File: kyo-core/shared/src/test/scala/kyoTest/envsTest.scala
Comment (hearnadam) [inline kyo-core/shared/src/test/scala/kyoTest/envsTest.scala:263]: Will followup to ensure we are not leaving unnecessary diff.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:59]: Should we call this `Layer` or `Layers`?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:57]: We can avoid casting if we switch to an encoding which automatically pulls the remaining types (`In`) from a pending `Envs[In]`

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:11]: can we have readable names for these instead?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:57]: I think `Layers` for now but we can re-consider the decision to use plurals

File: kyo-core/shared/src/test/scala/kyoTest/layersTest.scala
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyoTest/layersTest.scala:60]: Could you add more tests validating the behavior with more complex dependency graphs?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:35]: Do we need the `Flat` bypass here now that `TypeMap` has an evidence for it?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:48]: Can you move these extensions to a `Layer` companion object?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:16]: Kyo provides stronger guarantees regarding purity across transformations of a computation. If an effect in a layer initialization invokes its continuation multiple times, this map can break referential transparency. Could this be a recursive method instead using an immutable `Map`?

File: kyo-core/shared/src/test/scala/kyoTest/layersTest.scala
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyoTest/layersTest.scala:12]: Do we need to tests for cyclic dependencies? Or is that not possible with the encoding?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:8]: I was hoping we could have only two type parameters but it doesn't seem possible with memoization. I think it's ok.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:76]: Kyo normally has overloaded methods for up to four parameters. Can you add them? I think Scala 3 has some tuple magic that would allow us to accept any number of parameters but I haven't tried it yet.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:81]: You can also use `zip` here: `zip(Envs.get[A], Envs.get[B]).map(f)`

Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:81]: Nice! Didn't know about `zip`. `.map(f)` is not possible because `zip`  produces a tuple...

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:76]: Done.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:8]: It should be possible to encode it as:

```
Layer[S & Envs[In], Out]
```
If we can unpack the `Envs[A]`...

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:11]: Sure! Any suggestions?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:35]: Nope, I can use `summon` instead.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:15]: Kit and I tried a recursive encoding and weren't able to make it work. Would you prefer immutable map + atomic?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:7]: Interesting! That'd be great

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:10]: Maybe to/and/then?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:15]: I'll give it a try later today

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:47]: Do you mean just nesting the extensions? or converting to object methods? ie: `Layer.run(layer)`

Comment (kitlangton) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:47]: I think nesting under the companion object. That should ensure they're always resolvable w/ implicit search.

File: kyo-core/shared/src/test/scala/kyoTest/layersTest.scala
Comment (hearnadam) [reply kyo-core/shared/src/test/scala/kyoTest/layersTest.scala:5]: I don't think that makes much sense. I did add a test for when `In =:= Out`, which we don't want to break memoization.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala]: Done.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala]: `to/and/using`

Comment (kitlangton) [reply kyo-core/shared/src/main/scala/kyo/layers.scala]: Isn’t using a keyword? :P

Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala]: It compiles 😂

**Comment by calvinlfer** [comment]
Wow this is super nice! 
I'm curious about Envs now, with Layers, do you see Envs as a way to ask for things and make it appear as Layer requirements? or does standalone Envs still have its use?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:15]: Sorry for the delay, I took a look but I can't see a way to avoid mutability for memoization. What do you think about adding `IOs` suspensions for the parts of the code that interact with the mutable map? It'd address the referential transparency issue but `Layer.run` would depend on `IOs`. Also, don't we need a concurrent map here since layers might be initialized in parallel?

Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:15]: As it stands today, we don't use any par related operators. That is something we could add in the future, but not sure if the complexity is warranted now.

A dependency on IOs should be alright, it will just remove the ability to use layers as a general memoization technique... IMO it's not too big of a deal as it's pretty dependency injection specific

**Comment by hearnadam** [comment]
@fwbrasil @kitlangton I seem to have cracked the code on an encoding with just 2 type parameters. Let me know what you think!

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:68]: I'm very happy with the result here! Without the `Envs[B]` specialization, the layer should be transparent to the underlying Kyo computation.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala]: Well maybe not...

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:37]: Minor: it's possible to avoid the `map` transformation:

```scala
IOs {
    memoMap.compute(self, { (_, _) => result })
    result
}
```

By the way, couldn't we use `putIfAbsent`?

**Review by fwbrasil — APPROVED** [review]

**Comment by hearnadam** [comment]
@fwbrasil the layers API is complete. There is minor polish further required for the error messaging, but the API should be stable - so we can merge this and followup if you are okay with that!

File: kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala
Comment (hearnadam) [inline kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala:198]: @kitlangton should we make the whole graph/validated private? Is there any reason we can't?

File: kyo-core/shared/src/main/scala/kyo/package.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/package.scala:11]: Can we move the extension to `kyo-combinators` and provide a `Envs.runLayer`? Maybe we could make `Envs.runTypeMap` not public as well?

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:55]: do we need this method to be part of the public API? It feels like `Layers` is the proper API and `TypeMap` is more of a implementation detail. Also, please add the type annotation to `asInstanceOf`. It's possible the compiler ends up inferring `Nothing` in case we change this code.

File: kyo-core/shared/src/test/scala/kyoTest/layersTest.scala
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyoTest/layersTest.scala:186]: pretty cool!!

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:8]: Kyo's codebase favors `abstract class` over `trait` where possible to avoid interface dispatches, which have some additional performance overhead. Can you check if this could be `abstract class`? Also, please mark the methods as `final` so the JIT can more easily optimize invocations.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/layers.scala:93]: Kyo's naming typically use `init` for this kind of method. We can review to start using `make` but can you rename to `init` here for consistency for now?

File: kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala:76]: Indentation is off here.

File: kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala:151]: I should have a proper public impl for this but I think it's ok to keep it internal for now!

File: kyo-core/shared/src/test/scala/kyoTest/layersTest.scala
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyoTest/layersTest.scala:222]: Why does this fail? Is it because `TypeMap` is covariant?

File: kyo-core/shared/src/test/scala/kyoTest/layersTest.scala
Comment (fwbrasil) [inline kyo-core/shared/src/test/scala/kyoTest/layersTest.scala:115]: It's great how the API plays well with other effects 👏

File: kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala
Comment (fwbrasil) [inline kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala:36]: Can we have a better placeholder message for now? :)

File: kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala
Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/internal/LayerMacros.scala:198]: Agree! We could move impls to a companion object to avoid polluting the `kyo.internal` namespace as well.

File: kyo-core/shared/src/test/scala/kyoTest/layersTest.scala
Comment (hearnadam) [reply kyo-core/shared/src/test/scala/kyoTest/layersTest.scala:222]: The goal is that we should fail compilation when too many layers are provided. Currently, we happily only use some of the layers to construct a new layer.

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:8]: Done!

File: kyo-core/shared/src/main/scala/kyo/layers.scala
Comment (hearnadam) [reply kyo-core/shared/src/main/scala/kyo/layers.scala:55]: I went ahead and marked it private as I don't have a good example of using `.run` on a layer, but it may need to be public eventually.

**Comment by hearnadam** [comment]
@fwbrasil all feedback should be addressed. The implementation is sound, but I think we can revisit error reporting a bit.

Regarding naming, I followed your comments. I plan to revisit naming for all of Kyo once the new core is merged.

**Comment by fwbrasil** [comment]
woohoo! The next release will be amazing 🎉

**Pull request merged**
